app/multitenant/consul_pipe_router.go

```diff
@@ -0,0 +1,605 @@
+	_, err := pr.cas(key, func(p *consulPipe) (*consulPipe, bool, error) {
+		if p == nil {
+			p = &consulPipe{
+				CreatedAt: mtime.Now(),
+			}
+		}
+		if !p.DeletedAt.IsZero() {
+			return nil, false, fmt.Errorf("Pipe %s has been deleted", key)
+		}
+		end := p.end(e)
+		if end != "" && end != pr.advertise {
+			return nil, true, fmt.Errorf("Error: Pipe %s has existing connection to %s", key, end)
+		}
+		p.setEnd(e, pr.advertise)
+		p.incr(e)
+		return p, false, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Next wait for other end to connect
+	// at this point we 'own' one end
+	pipe, err := pr.watch(key, mtime.Now().Add(10*longPollDuration), func(p *consulPipe) (bool, error) {
+		return p.otherEnd(e) == "", nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return pipe, nil
+}
+
+func (pr *consulPipeRouter) Exists(ctx context.Context, id string) (bool, error) {
+	userID, err := pr.userIDer(ctx)
+	if err != nil {
+		return false, err
+	}
+	key := fmt.Sprintf("%s%s-%s", pr.prefix, userID, id)
+	consulPipe, err := pr.get(key)
+	if err != nil {
+		return false, err
+	}
+	return consulPipe == nil || consulPipe.DeletedAt.IsZero(), nil
+}
+
+func (pr *consulPipeRouter) Get(ctx context.Context, id string, e app.End) (xfer.Pipe, io.ReadWriter, error) {
+	userID, err := pr.userIDer(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	key := fmt.Sprintf("%s%s-%s", pr.prefix, userID, id)
+	log.Infof("Get %s:%s", key, e)
+
+	// first check we own the pipe in consul
+	_, err = pr.negotiate(key, e)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// next see if we already have a active pipe
+	pr.mtx.Lock()
+	pipe, ok := pr.pipes[key]
+	if !ok {
+		pipe = xfer.NewPipe()
+		pr.pipes[key] = pipe
+	}
+	pr.mtx.Unlock()
+
+	myEnd, _ := pipe.Ends()
+	if e == app.ProbeEnd {
+		_, myEnd = pipe.Ends()
+	}
+	return pipe, myEnd, nil
+}
+
+func (pr *consulPipeRouter) Release(ctx context.Context, id string, e app.End) error {
+	userID, err := pr.userIDer(ctx)
+	if err != nil {
+		return err
+	}
+	key := fmt.Sprintf("%s%s-%s", pr.prefix, userID, id)
+	log.Infof("Release %s:%s", key, e)
+
+	// atomically clear my end of the pipe in consul
+	_, err = pr.cas(key, func(p *consulPipe) (*consulPipe, bool, error) {
+		if p == nil {
+			return nil, false, fmt.Errorf("Pipe %s not found", id)
+		}
+		if p.end(e) != pr.advertise {
+			return nil, false, fmt.Errorf("Pipe %s not owned by us!", id)
+		}
+		refs := p.decr(e)
+		if refs == 0 {
+			p.setEnd(e, "")
+		}
+		return p, true, nil
+	})
+	return err
+}
+
+func (pr *consulPipeRouter) Delete(ctx context.Context, id string) error {
+	userID, err := pr.userIDer(ctx)
+	if err != nil {
+		return err
+	}
+	key := fmt.Sprintf("%s%s-%s", pr.prefix, userID, id)
+	log.Infof("Delete %s", key)
+
+	_, err = pr.cas(key, func(p *consulPipe) (*consulPipe, bool, error) {
+		if p == nil {
+			return nil, false, fmt.Errorf("Pipe %s not found", id)
+		}
+		p.DeletedAt = mtime.Now()
```

Comment:
I was thinking we might need a read lock around calls to `p.DeletedAt.IsZero()` (etc), since we're modifying the timestamp from multiple threads concurrently. Not sure now...